### PR TITLE
368 swaplist component (1/2)

### PR DESCRIPTION
**Pull request by vivz753**
**Explain the details for making this change. What existing problem does the pull request solve?**

Intialized the swaplist component with sortable mixin, integrated with ids-card and ids-list-view with multiselect functionality
- Refactored the select functionality from ids-list-builder --> ids-list-view
- Created multiselect option
- Fills the first list with the data
- Render the arrow buttons dynamically based on whether list is in the middle or not

TODO: 
- Decide how to manage data across list-views, keep them encapsulated with each data property of each list-view, or in addition to that add a data property on swaplist component that contains all data... maybe have `IdsSwapList.data = [[data1], [data2], [data3], ...]` where data1 is the first IdsListView.data
- Add button support for swapping elements in the DOM across the ids-list-views (do this first, then use that function when dragend is fired--see sub bullet)
  - Add dragging support for dragging elements horizontally across the ids-list-views
- Add keyboard support for shift click on swap list
- Add customizable title
- Write tests
- Add README.md
- Add Types

**Related github/jira issue (required)**:

- Addresses #368 but isn't finished yet. Passing this off to @jmacaluso711 

**Steps necessary to review your pull request (required)**:

- http://localhost:4300/ids-swaplist

Screen Shot 2021-12-27 at 8 58 51 AM

**Included in this Pull Request**:
- [ ] An e2e or functional test for the bug or feature.
- [ ] A note to the change log.

### CODE REVIEW

File: demos/ids-list-view/selectable.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{{> ../layouts/head.html }}
```
Comment (vivz753) [inline demos/ids-list-view/selectable.html:1]: To show case the new `selectable` functionality on `ids-list-view`

File: src/components/ids-swaplist/ids-swaplist.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    super();
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    this.defaultTemplate = `${this.querySelector('template')?.innerHTML || ''}`;
+    const l = this.container.querySelector('ids-list-view');
+    l.defaultTemplate = this.defaultTemplate;
+    l.data = [
```
Comment (vivz753) [inline src/components/ids-swaplist/ids-swaplist.js:35]: Dummy data for now, remove later

File: src/components/ids-swaplist/ids-swaplist.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  constructor() {
+    super();
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    this.defaultTemplate = `${this.querySelector('template')?.innerHTML || ''}`;
+    const l = this.container.querySelector('ids-list-view');
+    l.defaultTemplate = this.defaultTemplate;
```
Comment (vivz753) [inline src/components/ids-swaplist/ids-swaplist.js:34]: Need to override the list-view defaultTemplate property in order to grab the `<template>` format from HTML

**Comment by tmcconechy** [comment]
@vivz753 i tried to fix the treemap test here https://github.com/infor-design/enterprise-wc/pull/470/files#diff-1e7e6a76956341e0f77e665ec1393d00642775dcd7286ea57113259e0ce185c5L18 so you could copy the same change in this branch and try it. I think its intermittent so not 100% sure if thats going to totally solve it.

File: demos/ids-list-view/index.yaml [hunk 4]
```diff
@@ -13,6 +13,10 @@
     description: Showing a list view using virtual scroll
   - link: side-by-side.html
     description: Showing a list view side by side with a 4.x busy indicator
+  - link: sortable.html
+    description: Showing a list view with the sortable functionality
+  - link: selectable.html
```
Comment (tmcconechy) [inline demos/ids-list-view/index.yaml:18]: I just realized on the card and datagrid we called this `single-selection` and `multiple-selection` and the option is called `selection: 'single/multiple` so could be something to make consistent.

File: demos/ids-swaplist/index.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+---
+ component: SwapList 
+ link: ids-swaplist
+ description: Displays SwapList component 
+ category: Choose one -> [Charts and Visualizations, Layouts, Form Inputs, Navigation and Interaction, Lists, Messages and Alerts]
```
Comment (tmcconechy) [inline demos/ids-swaplist/index.yaml:5]:
```suggestion
 category: Lists
```

File: src/components/ids-card/ids-card.scss [hunk 6]
```diff
@@ -61,6 +61,10 @@
 
 ::slotted([slot='card-header']) {
   @include py-12();
+
+  display: flex;
```
Comment (tmcconechy) [inline src/components/ids-card/ids-card.scss:65]: I'd check all card examples if you didnt... can compare to https://main.wc.design.infor.com/ids-card/ 
I went through a couple and it looked ok but double checking...

File: src/core/ids-attributes.js [hunk 7]
```diff
@@ -156,6 +156,7 @@ export const attributes = {
   MODE: 'mode',
   MORE_ACTIONS: 'more-actions',
   MULTIPLE: 'multiple',
+  MULTISELECT: 'multiselect',
```
Comment (tmcconechy) [inline src/core/ids-attributes.js:159]: Again think we should use `selection: single | multiple` like cards and datagrid to be consistent and there is the start of a mixin we can use https://github.com/infor-design/enterprise-wc/pull/470 for the properties

File: test/ids-swaplist/ids-swaplist-func-test.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
+/**
+ * @jest-environment jsdom
+ */
+import IdsSwapList from '../../src/components/ids-swaplist/ids-swaplist';
+
+const HTMLSnippets = {
+  VANILLA_COMPONENT: (
```
Comment (tmcconechy) [inline test/ids-swaplist/ids-swaplist-func-test.js:7]:
```suggestion
  SWAPLIST_COMPONENT: (
```

Would this be more clear?

**Review by tmcconechy — COMMENTED** [review]
Just some quick things i saw. I wont take this too far if John is taking it over from here and its incomplete. (styling / ability to drag between zones)

Just need checks passing is all.

File: src/core/ids-attributes.js (hunk 7 above)
Comment (vivz753) [reply src/core/ids-attributes.js:159]: This is better than what I came up with, thanks. I've changed it.

File: demos/ids-list-view/index.yaml (hunk 4 above)
Comment (vivz753) [reply demos/ids-list-view/index.yaml:18]: The tree components use `SELECTABLE` and datagrid component uses `SELECTION`. I'm just going to leave it as `SELECTABLE` and let you guys decide what to make the standard

File: src/components/ids-card/ids-card.scss (hunk 6 above)
Comment (vivz753) [reply src/components/ids-card/ids-card.scss:65]: I was waiting to see the percy screenshot diffs to check this, will check again tmrw

**Comment by tmcconechy** [comment]
Not sure i understand what is failing in the tests from the output. But just wanted to mention you might see this https://github.com/infor-design/enterprise-wc/runs/4651115551?check_suite_focus=true#step:9:244 which is how i realized the snap shot was missing from being commited to repo. So basically it just makes a new one everytime and never compares it to the last one. So if you add a new snap shot test should commit the snap file.

File: src/components/ids-swaplist/ids-swaplist.d.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export default class IdsSwapList extends HTMLElement {
+
```
Comment (vivz753) [inline src/components/ids-swaplist/ids-swaplist.d.ts:2]: Add count: number

**Comment by vivz753** [comment]
Fixed tests!

**Review by tmcconechy — APPROVED** [review]
Looks good for 1/2

**Comment by tmcconechy** [comment]
@jmacaluso711 do you need this merged to work on it? If so please approve.

**Review by jmacaluso711 — APPROVED** [review]
@tmcconechy yes, let's get this merged. Approved

**Pull request merged**